kratos/processes/multipoint_constraint_to_element_process.cpp

```diff
@@ -0,0 +1,343 @@
+        {
+            return (Left.first == Right.first && Left.second == Right.second)
+                || (Left.first == Right.second && Left.second == Right.first);
+        }
+
+        first_type first;
+
+        second_type second;
+    };
+
+
+    /// @brief Pair of symmetric ID pairs.
+    /// @note This pair is NOT symmetric, but the subpairs are.
+    template <class TFirst, class TSecond>
+    struct NestedIdPair
+    {
+        using first_type = SymmetricIdPair<TFirst>;
+
+        using second_type = SymmetricIdPair<TSecond>;
+
+        struct hash
+        {
+            std::size_t operator()(NestedIdPair Instance) const noexcept
+            {
+                auto first_hash = typename first_type::hash()(Instance.first);
+                const auto second_hash = typename second_type::hash()(Instance.second);
+                HashCombine(first_hash, second_hash); // <== mutates first_hash
+                return first_hash;
+            }
+        }; // struct hash
+
+        friend bool operator==(NestedIdPair Left, NestedIdPair Right) noexcept
+        {
+            return Left.first == Right.first && Left.second == Right.second;
+        }
+
+        first_type first;
+
+        second_type second;
+    };
+
+    std::optional<ModelPart*> mpInputModelPart;
+
+    std::optional<ModelPart*> mpOutputModelPart;
+
+    std::optional<const Variable<double>*> mpVariable;
+
+    std::optional<IntervalUtility> mInterval;
+
+    /// Map associating each constrained variable pair with a sub model part
+    /// in the output model part.
+    std::unordered_map<
+        SymmetricIdPair<Variable<double>::KeyType>,
+        ModelPart*,
+        SymmetricIdPair<Variable<double>::KeyType>::hash
+    > mSubModelPartMap;
+
+    /// Map associating an @ref MasterSlaveConstraint "MPC" and an element
+    /// with each pair of nodes and variables.
+    std::unordered_map<
+        NestedIdPair<Node::IndexType,Variable<double>::KeyType>,
+        std::pair<
+            Element::IndexType,
+            MasterSlaveConstraint::IndexType
+        >,
+        NestedIdPair<Node::IndexType,Variable<double>::KeyType>::hash
+    > mConstraintMap;
+
+    /// @brief Find the largest @ref Element ID in a @ref ModelPart.
+    static Element::IndexType GetLargestElementId(const ModelPart& rModelPart)
+    {
+        // IDs are 1-based
+        Element::IndexType output = 1;
+
+        // Elements are stored in an ordered container, sorted by their IDs,
+        // so the last element in the container should have the largest ID.
+        if (!rModelPart.Elements().empty()) {
+            output = rModelPart.Elements().back().Id();
+        }
+
+        // Max reduce across MPI ranks
+        const DataCommunicator& r_mpi = rModelPart.GetCommunicator().GetDataCommunicator();
+        return static_cast<Element::IndexType>(r_mpi.MaxAll(static_cast<long unsigned int>(output))); // <== explicit casts for MSVC
```

Comment:
I'll need to think this through because each rank will start issuing element IDs at the same index.